### ISSUE TITLE
User attributes mapping and role mapping with third-party authorization (e.g. SAML)

### ISSUE THREAD

**Issue by dominikklein** [zammad/zammad#4943]
### What

The user attributes are only mapped once (and only special attributes) during the usage of the third-party authorization. 
Because of that, it's for example currently not possible to use only an SAML authorization when you want to have the user attributes up to date (and also sync more than the "simple" defined attributes) without an additional LDAP sync in place. Also, it's not possible to sync any role assignment.

### Use cases

As a company, I want to have one central place for all my user handling without the need to connect one part for authentication and one part for a general user sync.

### Intended Outcome

The user attributes should be synced during every third-party login (the "external source" should be the central data point). Also, it should be possible to handle a role and group mapping (currently this will only be related to SAML).
With this, it should be possible in the future to use SAML without the need for an LDAP sync (=this means not it's no longer allowed/possible when it's still needed).

#### Restrictions
- No creation of new roles
- No creation of new groups

### Technical implementation

For every third-party authorization, we should implement the needed mapping for the known attributes. Especially for SAML, it should be possible to sync every attribute when the mapping on the IDP side provides the attribute in the "needed" way.

In the SAML context also roles should be mapped. For this, the given role needs already be present inside the Zammad system.

**In the current user attribute mapping the usage of the "nickname" attribute should be removed, because it's no valid uid for the login.**

**Comment by samip5** [comment]
Can someone please share some light into the current situation with this? With no OpenID Connect generic support, and the role thing, this is quite a pain to use when an org doesn't want to manage users in multiple different places (eg. an org without LDAP as an example).

**Comment by noriakisana** [comment]
Any updates on this?

**Comment by samip5** [comment]
> Any updates on this?

Workaround is eg Authentik with LDAP outpost.

**Comment by noriakisana** [comment]
yeah, we wanted to stop the process, where each application has the credentials of every user :D

**Comment by MrGeneration** [comment]
> application has the credentials of every user

The password is not synced at all and always checked live against LDAP. If you have further questions regarding this topic, feel free to spawn in our community: https://community.zammad.org.

**Comment by quentinmit** [comment]
I also observed this bug. Zammad does not appear to update a user's name/email/etc. based on a SAML login if the user already exists in Zammad. Every login should sync any provided attributes to the user. Also, there is a feature request to support assigning roles based on a SAML attribute.

**Comment by ldallolio** [comment]
Hi, 
Same here, here is our use case : in our (very big) organisation the LDAP is huge and we cannot (not even allowed and it would take hours) sync it fully, so we only sync some parts of it for agents (to assign roles) and not for customers. 
Customers can still connect via SAML (easier for them with respect to ldap, also better for security reason since the application will never see the password during authentication).
The only missing feature is, as specified in this ticket, that customers creating their accounts by SAML will have no attribute specified, even if the SAML would provide attribute metadata in the response. This impacts triggers, reporting that would react to user attributes.

**Comment by seal61** [comment]
any updates on this with the new zammad 6.5 release?

**Comment by Thiwico** [comment]
I would also like to have role-based access via OpenID and automatic assignment of groups in Zammad.

**Comment by jlegue** [comment]
This is an important function for any SSO implementation. Assigning a user to both groups and roles manually after their first SSO login defeats the purpose of single sign on where it should all be manage in one place. Add the user to IDP groups to be assiged to application groups/roles.

**Comment by quentinmit** [comment]
We applied a local patch to sync SAML groups to Zammad roles: https://github.com/mitmh2025/hunt2025/blob/main/infra/pkgs/zammad/saml-roles.patch

It might give y'all a starting point if you are willing to patch your local installation as well.

**Comment by noriakisana** [comment]
for some people maybe a useful workaround (for us the purpose of sso login is that our people gets a "global" unified login and protect the from fishing and co.)

(its double stuff), currently we connected zammad with the ldap outpost (via authentik) and added it also as sso to zammad, yeah it is very hacky but it works, since it get double checked (but initially can lead into some problem because of the SAML claim (if i remember it correctly))

I think long term this should be implemented

**Comment by yuhongwei380** [comment]
which version support this saml feature?

**Comment by noriakisana** [comment]
The current should i can See how Long we already have this implemented, If needed

**Comment by yuhongwei380** [comment]
> The current should i can See how Long we already have this implemented, If needed目前我能看看我们已经实现多久了，如果需要的话

thanks for do it ! we use the docker-compose zammad repo ;

**Comment by noriakisana** [comment]
We have it for over 2 years now so at least the zammad docker compose Release v12.0.0 should do it.

**Comment by yuhongwei380** [comment]
> We have it for over 2 years now so at least the zammad docker compose Release v12.0.0 should do it.我们已经用了两年多了，至少 zammad docker compose 版本 v12.0.0 应该能做到。

thanks for search it ,we use the 6.5.0 version(v13.4.0) ,does this feature had any docs ?  I remember i had test it ,but    I don‘t know how to config role( connect to keycloak and the user had roles ) for this configration for zammad

<img width="774" height="94" alt="Image" src="https://github.com/user-attachments/assets/51916496-0586-4daf-b534-0f6bca42cc88" />

**Comment by noriakisana** [comment]
Maybe Something Like this?

https://admin-docs.zammad.org/en/latest/settings/security/third-party/saml/saml-keycloak.html

**Comment by yuhongwei380** [comment]
> Maybe Something Like this?
> 
> https://admin-docs.zammad.org/en/latest/settings/security/third-party/saml/saml-keycloak.html

thanks, i will try it with saml , and test the role ; now i used with OIDC between keycloak and zammad

**Comment by Dreamzallsi2** [comment]
> > Maybe Something Like this?
> > https://admin-docs.zammad.org/en/latest/settings/security/third-party/saml/saml-keycloak.html
> 
> thanks, i will try it with saml , and test the role ; now i used with OIDC between keycloak and zammad

I am looking for solution how to applied role into my zammad via OpenID / SAML 
Do you have solution yet ?? or only saml patch that person have talked about ??

**Comment by yuhongwei380** [comment]
> > > Maybe Something Like this?
> > > https://admin-docs.zammad.org/en/latest/settings/security/third-party/saml/saml-keycloak.html
> > 
> > 
> > thanks, i will try it with saml , and test the role ; now i used with OIDC between keycloak and zammad
> 
> I am looking for solution how to applied role into my zammad via OpenID / SAML Do you have solution yet ?? or only saml patch that person have talked about ??

so sry for that ,i am focus on my job  , no time to test this

**Comment by biowilli** [comment]
single source of truth for identities with OIDC would be really cool!  <3

**Comment by jschneider97** [comment]
Are there any Updates on whether/when this will be implemented? We would really love to see this feature!

**Comment by bokkabonga** [comment]
We really hope this will be implemented soon.

**Comment by RaitoBezarius** [comment]
Hi there @dominikklein; I'd be willing to send a PR to add this feature, is there any specific etiquette to respect to make this happen? I'm mostly interested about OIDC FWIW.

**Comment by ClFeSc** [comment]
@RaitoBezarius Just for your information: We've already offered a PR for this back in 2022: #4211

That PR was then closed in favour of this Issue.

**Comment by RaitoBezarius** [comment]
@ClFeSc Thank you for the mention. 4 years have passed since then and it seems like that the Zammad team was not able to get it, perhaps @dominikklein can revisit and steer the external contributors to achieve a vision they would consent to merge?
